decode.go

```diff
@@ -102,40 +102,22 @@ func (p *parser) peek() libyaml.EventType {
 	// It's curious choice from the underlying API to generally return a
 	// positive result on success, but on this case return true in an error
 	// scenario. This was the source of bugs in the past (issue #666).
-	if !p.parser.Parse(&p.event) || p.parser.ErrorType != libyaml.NO_ERROR {
-		p.fail()
+	if err := p.parser.Parse(&p.event); err != nil {
+		p.fail(err)
 	}
 	return p.event.Type
 }
 
-func (p *parser) fail() {
-	var line int
-	if p.parser.ContextMark.Line != 0 {
-		line = p.parser.ContextMark.Line
-		// Scanner errors don't iterate line before returning error
-		if p.parser.ErrorType == libyaml.SCANNER_ERROR {
-			line++
+func (p *parser) fail(err error) {
+	var parserErr *libyaml.ParserError
+	if errors.As(err, &parserErr) {
+		err = &ParserError{
+			Message: parserErr.Message,
+			Line:    parserErr.Line,
+			Column:  parserErr.Column,
 		}
```

Comment:
Why do you convert from internal ParserError to an external one? Anyway it will be converted to yamlError and then panicked.

I think you can get rid of `fail` method completely and use `fail` function, because you already removed all logic